Lib/typing.py

```diff
@@ -194,23 +201,82 @@ def _collect_type_vars(types):
     """
     tvars = []
     for t in types:
-        if isinstance(t, _TypeVarLike) and t not in tvars:
+        if isinstance(t, _TypeVarTypes) and t not in tvars:
             tvars.append(t)
         if isinstance(t, (_GenericAlias, GenericAlias)):
             tvars.extend([t for t in t.__parameters__ if t not in tvars])
     return tuple(tvars)
 
 
-def _check_generic(cls, parameters, elen):
-    """Check correct count for parameters of a generic cls (internal helper).
-    This gives a nice error message in case of count mismatch.
+def _check_type_parameter_count(cls,
+                                type_params, expected_num_type_params=None):
+    """Checks for correct count of type parameters of a generic class.
+
+    Gives a nice error message in case of count mismatch.
+
+    Args:
+        cls: The generic class in question.
+        type_parameters: A tuple of the type parameters the user is attempting
+          to use with to the generic class. For example, for 'Tuple[int, str]',
+          `type_parameters` would be `(int, str)`.
+        expected_num_type_params: The exact number of type parameters expected.
+          If `None`, this is determined by inspecting `cls.__parameters__`.
     """
-    if not elen:
-        raise TypeError(f"{cls} is not a generic class")
-    alen = len(parameters)
-    if alen != elen:
-        raise TypeError(f"Too {'many' if alen > elen else 'few'} parameters for {cls};"
-                        f" actual {alen}, expected {elen}")
+    actual_num_type_params = len(type_params)
+
+    # These should have been checked elsewhere.
+    assert not any(isinstance(p, TypeVarTuple) for p in type_params)
+    if expected_num_type_params is not None:
+        assert expected_num_type_params > 0
```

Comment:
Could you address the comment later in the file?

```
# _nparams is the number of accepted parameters, e.g. 0 for Hashable,
# 1 for List and 2 for Dict.  It may be -1 if variable number of
# parameters are accepted (needs custom __getitem__).
```